integration/v3_grpc_test.go

```diff
@@ -193,3 +196,157 @@ func TestV3DeleteRange(t *testing.T) {
+		{
+			[]string{},
+			"bar",
+			[]string{"foolong"},
+			"bar",
+			[]pb.WatchRequest{{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foo")}}},
+
+			1,
+			[]storagepb.Event{
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("foolong"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+			},
+		},
+		// watch the prefix, non-matching
+		{
+			[]string{},
+			"bar",
+			[]string{"foo"},
+			"bar",
+			[]pb.WatchRequest{{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foobar")}}},
+
+			0,
+			[]storagepb.Event{},
+		},
+		// watch the prefix, matching, unsynced, multiple-events
+		{
+			[]string{"a", "aa", "aaa"},
+			"bar",
+			[]string{"a", "aa", "aaa"},
+			"barbar",
+			[]pb.WatchRequest{
+				{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("a"), StartRevision: 1}},
+			},
+
+			3,
+			[]storagepb.Event{
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("a"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("aa"), Value: []byte("bar"), CreateRevision: 2, ModRevision: 2, Version: 1},
+				},
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("aaa"), Value: []byte("bar"), CreateRevision: 3, ModRevision: 3, Version: 1},
+				},
+			},
+		},
+	}
+
+	for i, tt := range tests {
+		clus := newClusterGRPC(t, &clusterConfig{size: 3})
+		kvc := pb.NewKVClient(clus.RandConn())
+
+		for _, k := range tt.initialKeys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte(tt.initialValue)}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Errorf("#%d-1: couldn't put key (%v)", i, err)
+			}
+		}
+
+		wAPI := pb.NewWatchClient(clus.RandConn())
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Errorf("#%d: wAPI.Watch error: %v", i, err)
+		}
+
+		for _, r := range tt.watchRequests {
+			if err := wStream.Send(&r); err != nil {
+				t.Errorf("#%d: wStream.Send error: %v", i, err)
+			}
+		}
+
+		for _, k := range tt.keys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte(tt.value)}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Errorf("#%d-2: couldn't put key (%v)", i, err)
+			}
+		}
+
+		if len(tt.initialKeys) > 0 {
+			// give time to run syncWatchers
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		cnt := 0
+		for cnt != tt.eventNum {
+			resp, err := wStream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				t.Errorf("#%d: wStream.Recv error: %v", i, err)
+			}
+
+			switch {
+			case resp.Created:
+				t.Logf("#%d: watcher created: id %08x\n", i, resp.WatchId)
+			default:
+				if tt.eventNum != len(resp.Events) {
+					t.Errorf("#%d: len(resp.Events) got = %d, want = %d", i, len(resp.Events), tt.eventNum)
+				}
+
+				copied := []storagepb.Event{}
+				for _, r := range resp.Events {
+					if r != nil && r.Kv != nil {
+						copied = append(copied, *r)
+						cnt++
+					}
+				}
+
+				if !reflect.DeepEqual(copied, tt.events) {
```

Comment:
if all the events don't appear in a single response, then shouldn't this check go outside the `for` loop?